* TODO Genomics
** DONE Min size of sequences set to 6bp, optional argument
   CLOSED: [2022-01-10 Mon 10:20]
** TODO Implement ambiguous bases in GC/CpG content
** TODO Test for GC/CpG ambiguous bases
** DONE Test for GC/CpG N nucleotides
   CLOSED: [2022-01-10 Mon 14:32]
** DONE CpG, take care of N nucleotides
   CLOSED: [2022-01-10 Mon 14:32]
** Codon usage bias
** TODO Parse introns
** TODO Parse UTR (i.e. exon - CDS)

* TODO Pop statistics
** Pi, correct implementation, see https://github.com/ksamuk/pixy & https://github.com/RILAB/mop
** Substitution rate/proportion
** Get synonymous/non-synonymous sites
** mu, local mutation rate, estimated by the four-fold synonymous substitution rate
** PiN/PiS (see https://github.com/chasewnelson/SNPGenie)
** dN/dS
** Fst
** Dxy

* TODO Interface
** --header option if column names in the windows file (remove first row)
** --poplist "poplist" file with individuals/samples and associated population
** DONE Treat GFF as a windows file to augment with statistics: same column names
   CLOSED: [2022-01-10 Mon 16:43]
   
* TODO Optimization
** Optimization of GFF parsing
** Logs
** Parse gff if no id/parent attribute
** Validate functions for input data (e.g. check type and dimension)
** TODO Parallel processing in pislice (mapply)
** Splicing variants strategy (merge/keep, i.e. count as independent genes)
** Handle score in gff
** iPython Notebook for a tutorial of PiSlice
** Gff, parse header comments in a metadata accessor
** Vcf, parse header comments in a metadata accessor

* DONE
** DONE GC
   CLOSED: [2022-01-10 Mon 10:23]
** DONE GC123
   CLOSED: [2022-01-10 Mon 10:23]
** DONE Gene count
   CLOSED: [2022-01-10 Mon 10:23]
** DONE CpG proportion
   CLOSED: [2022-01-10 Mon 10:23]

