* TODO Genomics
** DONE Min size of sequences set to 6bp, optional argument
   CLOSED: [2022-01-10 Mon 10:20]
** TODO Implement ambiguous bases in GC/CpG content
** TODO Test for GC/CpG ambiguous bases
** DONE Gene length - bp
CLOSED: [2022-02-13 Sun 09:34]
** DONE Gene length - Number of exons
CLOSED: [2022-02-13 Sun 09:34]
** DONE Gene density
CLOSED: [2022-02-13 Sun 09:34]
** DONE Test for GC/CpG N nucleotides
   CLOSED: [2022-01-10 Mon 14:32]
** DONE CpG, take care of N nucleotides
   CLOSED: [2022-01-10 Mon 14:32]
** Codon usage bias
** TODO --seq option return the sequence sampled (beware, dataset set can be huge) -- mainly debug purpose or specific usage, otherwise use standard tools (vcftools, bedtools...)
** TODO gff.children get second order children (i.e. children of children)
** DONE GC non coding in windows
   CLOSED: [2022-01-12 Wed 10:12]
** DONE GC intergenic in windows
   CLOSED: [2022-01-12 Wed 10:18]
** DONE GC introns
   CLOSED: [2022-01-12 Wed 10:48]
** DONE Proportions of non-coding/intergenic in windows
   CLOSED: [2022-01-12 Wed 10:48]
** DONE Proportion of CDS in windows
CLOSED: [2022-02-13 Sun 10:08]
   ** TEST IT
   ** Splicing variants strategy?
   ** Merge overlaps
** DONE GC123 returns 'nan' when not gene or CDS
   CLOSED: [2022-01-11 Tue 19:04]
** DONE Parse introns
   CLOSED: [2022-01-11 Tue 18:51]
** TODO Parse UTR (i.e. exon - CDS)
** TODO How to consider splicing variants?
   ** Different strategies - Merging, keep the best
   ** best score
   ** longest
** TODO Filter gff per score, GT score
** Manage introns in UTR
** TODO GC-ratio (A+T)/(G+C)



* TODO Testing
** TODO Create a fake testing dataset
** implement the set of tests
   
* TODO Pop statistics
** Pi, correct implementation, see https://github.com/ksamuk/pixy & https://github.com/RILAB/mop
** Substitution rate/proportion
** Get synonymous/non-synonymous sites
** mu, local mutation rate, estimated by the four-fold synonymous substitution rate
** PiN/PiS (see https://github.com/chasewnelson/SNPGenie)
** dN/dS
** Tajima's D
** Fst
** Dxy

* TODO Interface
** --header option if column names in the windows file (remove first row)
** --poplist "poplist" file with individuals/samples and associated population
** DONE Treat GFF as a windows file to augment with statistics: same column names
   CLOSED: [2022-01-10 Mon 16:43]
** Progress bars

   
* TODO Optimization
** TODO Optimization of GFF parsing
** Logs
** Parse gff if no id/parent attribute
** Validate functions for input data (e.g. check type and dimension)
** TODO Parallel processing in pislice (mapply)
** Splicing variants strategy (merge/keep, i.e. count as independent genes)
** Handle score in gff - filtering
** iPython Notebook for a tutorial of PiSlice
** Gff, parse header comments in a metadata accessor
** Vcf, parse header comments in a metadata accessor

* Bugfix
** DONE gff.children only returns the first child, not children of children
   CLOSED: [2022-01-11 Tue 18:51]
   TODO Tests


   
* DONE
** DONE GC
   CLOSED: [2022-01-10 Mon 10:23]
** DONE GC123
   CLOSED: [2022-01-10 Mon 10:23]
** DONE Gene count
   CLOSED: [2022-01-10 Mon 10:23]
** DONE CpG proportion
   CLOSED: [2022-01-10 Mon 10:23]

